docs: add blob envelope generator capability (#326)
notaryproject/notaryproject: specs/plugin-extensibility.md

*supported-contract-versions* - The list of contract versions supported by the plugin. Currently this list must include only one version, per major version. Post initial release, Notation may add new features through plugins, in the form of new commands (e.g. tsa-sign for timestamping), or additional request and response parameters. Notation will publish updates to plugin interface along with appropriate contract version update. Backwards compatible changes (changes for which older version of plugin continue to work with versions of Notation using newer contract version) like new optional parameters on existing contracts, and new commands will be supported through minor version contract updates, breaking changes through major version updates. To maintain forward compatibility plugin implementors MUST ignore unrecognized attributes in command request which are introduced in minor version updates of the plugin contract. Plugin `get-plugin-metadata` command returns the contract version a plugin supports. Notation will evaluate the minimum plugin version required to satisfy a user's request, and reject the request if the plugin does not support the required version.

> *capabilities* - A non-empty list of features supported by the plugin. Each capability requires the plugin to implement specific commands. Notation will evaluate the required capability for a user's request and will reject the request if the plugin does not support it.

Author: Updated.

> Valid values and required commands:

Author: Updated.

  - `SIGNATURE_GENERATOR.RAW`: Generates a raw signature for OCI or blob payloads.
    - Required commands: `describe-key`, `generate-signature`
  - `SIGNATURE_GENERATOR.ENVELOPE`: Generates a complete signature envelope for OCI payloads.
    - Required command: `generate-envelope`

>   - `SIGNATURE_GENERATOR.BLOB_ENVELOPE`: Generates a complete signature envelope for blob (local file) payloads.

Reviewer: `BLOB_ENVELOPE` or `ENVELOPE_FOR_BLOB`?

Reviewer: Which is better?

Author: Updated to `ENVELOPE_FOR_BLOB`

>   - `SIGNATURE_GENERATOR.ENVELOPE_FOR_BLOB`: Generates a complete signature envelope for blob (local file) payloads.

Reviewer 2: maybe we can remove `(local file)`, because it can be a blob stored in a storage in the future, then it is not a file, instead could be a reference to blob in the storage?

Author: Updated.

### Signature Generator

> This interface targets plugins that integrate with providers of basic cryptographic operations (e.g., Local PIV/PKCS#11 hardware tokens, remote KMS, or key vault services). Plugins that target this interface will only generate a raw signature given a payload to sign. Notation will package this signature into a signature envelope, and generate the signature manifest or signature file. Notation will also generate the TSA signature if indicated by the user. The plugin does not need to be signature envelope format aware, and will continue to work if Notary Project adopts additional signature formats.

Author: Updated.

> This interface targets plugins that integrate with providers of basic cryptographic operations (e.g., Local PIV/PKCS#11 hardware tokens, remote KMS, or key vault services). Plugins that target this interface will only generate a raw signature given a payload to sign. Implementations of the Notary Project spec is responsible for packaging this signature into a signature envelope along with a signature manifest (OCI) or a signature file (blob). Implementations of the Notary Project spec is also responsible for generating the TSA signature if indicated by the user. The plugin does not need to be signature envelope format aware and will continue to work if Notary Project adopts additional signature formats.

Reviewer 3: ```suggestion
This interface targets plugins that integrate with providers of basic cryptographic operations (e.g., Local PIV/PKCS#11 hardware tokens, remote KMS, or key vault services). Plugins that target this interface will only generate a raw signature given a payload to sign. Plugin developers who implement the Notary Project specification are responsible for packaging this signature into a signature envelope along with a signature manifest (OCI) or a signature file (blob). Plugin developers are also responsible for generating the TSA signature if indicated by the user. The plugin does not need to be signature envelope format aware and will continue to work if Notary Project adopts additional signature formats.
```

Author: Updated.

    2. Else if, plugin supports capability `SIGNATURE_GENERATOR.ENVELOPE` *(covered in next section)*
    3. Return an error

**Blob signing:**
1. Given a user request to sign a local file (blob), with signing key `keyName`
2. Determine if the registered key uses a plugin
3. Execute the plugin with the `get-plugin-metadata` command and check its capabilities:
    1. **If the plugin supports `SIGNATURE_GENERATOR.RAW`:**
        1. Execute the plugin with the `describe-key` command, setting `request.keyId` and the optional `request.pluginConfig` to the values associated with signing key `keyName` in `config.json`.
        2. Generate the digest of the local file using the **hash algorithm** specified in the `keySpec` from the `describe-key` response. Construct the payload to be signed using this digest, as described in the [signature specification](signature-specification.md#payload).
        3. Follow the same signing steps as in the OCI signing workflow above (i.e., generate the signature, validate it, and assemble the signature envelope).
        4. Generate the signature file for the resulting signature envelope.

>     2. **Else if** plugin supports capability `SIGNATURE_GENERATOR.ENVELOPE` *(covered in next section)*

Reviewer: `BLOB_ENVELOPE`?

Author: Nice catch. Updated.

>     2. **Else if** plugin supports capability `SIGNATURE_GENERATOR.ENVELOPE` *(covered in next section)*

Author: Nice catch. Updated.
